auction/src/lib.rs

```diff
@@ -610,35 +670,49 @@ impl<T: Config> Pallet<T> {
 		new_owner: &T::AccountId,
 		price: BalanceOf<T>,
 		balance_source: Option<T::AccountId>,
-	) -> DispatchResult {
-		// Handle marketplace fees
+	) -> Result<(BalanceOf<T>, BalanceOf<T>), DispatchError> {
+		// Handle marketplace fees.
 		let marketplace = T::MarketplaceExt::get_marketplace(auction.marketplace_id)
 			.ok_or(Error::<T>::MarketplaceNotFound)?;
+		let mut nft = T::NFTExt::get_nft(nft_id).ok_or(Error::<T>::NFTNotFound)?;
 
-		let to_marketplace =
-			price.saturating_mul(marketplace.commission_fee.into()) / 100u32.into();
-		let to_auctioneer = price.saturating_sub(to_marketplace);
+		let mut commission_fee_amount: BalanceOf<T> = 0u32.into();
+		if let Some(commission_fee) = &marketplace.commission_fee {
+			match *commission_fee {
+				CompoundFee::Flat(x) => commission_fee_amount = x,
+				CompoundFee::Percentage(x) => commission_fee_amount = x * price,
+			};
+		}
+		let to_marketplace = commission_fee_amount;
+		let to_creator = nft.royalty * price.saturating_sub(to_marketplace);
+		let to_auctioneer = price.saturating_sub(to_marketplace).saturating_sub(to_creator);
 
-		let existence = if balance_source.is_none() { KeepAlive } else { AllowDeath };
+		let existence = if balance_source.is_some() { KeepAlive } else { AllowDeath };
 		let balance_source = balance_source.unwrap_or_else(|| Self::account_id());
 
-		// Transfer fee to marketplace
+		// Transfer fee to marketplace.
 		T::Currency::transfer(&balance_source, &marketplace.owner, to_marketplace, existence)?;
 
-		// Transfer remaining to auction creator
+		// Transfer fee to creator.
+		T::Currency::transfer(&balance_source, &nft.creator, to_creator, existence)?;
+
+		// Transfer remaining to auctionneer.
 		T::Currency::transfer(&balance_source, &auction.creator, to_auctioneer, existence)?;
 
-		T::NFTExt::set_owner(nft_id, new_owner)?;
-		T::NFTExt::set_listed_for_sale(nft_id, false)?;
+		nft.owner = new_owner.clone();
+		nft.state.is_listed = false;
+		T::NFTExt::set_nft(nft_id, nft)?;
```

Comment:
I know that this was here before but I don't like it now. It seems that it could cause problems because I don't expect that those kind of methods to modify NFT storage. What do you think that you move this out of this method?

This would mean that instead of passing an `NFT_ID`  as a parameter you would pass NFTData.